### ISSUE TITLE
sample is very slow on a big csv

### ISSUE THREAD

**Issue by lisbakke** [BurntSushi/xsv#119]
I'm sure this is expected, but I'm trying to sample a few thousand rows from a 10gb CSV and it takes minutes to complete. Is there a better way I should be doing this?

Thanks for the awesome tool.

**Comment by michelboaventura** [comment]
Did you try to index the database? For me it always greatly increases speed of almost every command.

**Comment by BurntSushi** [comment]
This isn't actionable. Please provide a reproducible example. You could at least list the specs of your hardware and the times of running other commands.

**Comment by lisbakke** [comment]
Mid 2014 macbook pro 15" running 10.13.4
2.5GHz core i7
16gb RAM

`xsv search -s cliff_value "[0]" noncliffs.csv` takes 3 mins 4 seconds

`xsv sample 129758 noncliffs.csv` takes 3 mins 13 seconds

**Comment by BurntSushi** [comment]
OK, so this has nothing to do with `xsv sample`.

@lisbakke You're not making this easy to debug. There's no reproducible example. I can't debug performance problems if you aren't willing to put in the work to provide as much data as possible. For example, it's conceivable that `xsv` is actually hitting disk on a file that big, but you don't actually bother to tell me what your storage hardware is. You should also be comparing `xsv` with other tools to get a decent baseline. e.g., How fast is `ggrep '[0]' noncliffs.csv`?

**Comment by dannguyen** [comment]
Hello, I know this issue has been closed, but just in case anyone is wondering, I have a real-world benchmark involving getting a 10% sample from a 6.5M row database. 

**tl;dr**: the average `time`, over 25 iterations, was **6.8 seconds**. Which, for my amateur purposes, is good enough :)

**My machine and OS:** [MacBook Pro (Retina, 15-inch, Late 2013) with 2.3GHz Intel Core I7 and 16GB RAM.](https://support.apple.com/kb/SP690?locale=en_US)

**OS**: Mac OS 10.13.5 High Sierra

**SSD:** I wrote to and read the data from an external USB3 SSD: The [Samsung T3 500GB](https://www.samsung.com/semiconductor/minisite/ssd/product/portable/t3/).


--------------

## Details

The data comes from the [city of Chicago's database of crime records, from 2001 to 2017](https://data.cityofchicago.org/Public-Safety/Chicago-Crimes-2001-through-2017/sgiv-a9r2/data).

The dataset's dimensions are **21** columns by **6.5 million** rows for a total of **1.33 GB**.

Bash commands to download and save the raw data, and then perform 25 iterations of `xsv sample` to produce a 10% (650,000 row) sample:


```sh
#### Download to working directory
curl -o crimes.csv \
     'https://data.cityofchicago.org/api/views/sgiv-a9r2/rows.csv?accessType=DOWNLOAD'

#### Run 25 times
time for i in {1..25}; do
     xsv sample 650000 crimes.csv > /dev/null
done
```

### Benchmark result over 25 times

The result output from `time`:

```
real    2m50.565s
user    2m10.141s
sys     0m40.210s
```

`(2 * 60 + 50) / 25` gets us: **6.8 seconds**

**Comment by BurntSushi** [comment]
@dannguyen Thanks for reporting that! That time feels reasonableish to me. The times could go up dramatically if `xsv` is hitting spinning disk.
